(Feature) Current exchanged value presented on invoices

sfount opened the pull request:
This pull requests adds a value in the selected currency to patient invoices. It uses the valid exchange rate at the time of invoice. 

If the enterprise currency is selected no exchange information is displayed.

**Receipt currency selection (Invoice registry module)**
![screenshot 2016-11-18 at 12 43 02](https://cloud.githubusercontent.com/assets/2844572/20429358/8cf8ef34-ad8e-11e6-8ddc-601d5db33752.png)

**Receipt displaying converted $:FC amount**
![screenshot 2016-11-18 at 12 43 27](https://cloud.githubusercontent.com/assets/2844572/20429423/cbcdd788-ad8e-11e6-8a87-8cb857ee7c1d.png)

The primary functionality is included in this pull request. Ideally to complete: 
- [x] Wrap currency selection from patient invoice registry module in a component (handles downloading of currencies etc.) 
- ~~Add receipt currency selection component to patient invoice module to allow receipts to be fetched in FC as they are made~~ To be designed

Future questions
* What is the best user experience for selecting the currency of an invoice or bill
* Should the entire invoice be displayed as FC - as the conversion happens on the lump sum (the total cost of the invoice) this could result in very inaccurate totals without hacks to prevent this

Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/888
Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/902

 - [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

sfount committed:
Receipt currency component/ service infrastructure

This commit adds required components and client side service
infrastructure to specifiy which currency an invoice should be made to a
patient in.

sfount committed:
Add exchange rate context for invoice receipts

This commit updates the template and the controller for adding exchange
rate values and an exchanged total to receipts.

jniles commented on client/src/partials/patient_invoice/registry/registry.js:
```diff
@@ -66,6 +82,11 @@ function InvoiceRegistryController(Invoices, bhConstants, Notify, Session, util,
     vm.loading = !vm.loading;
   }
 
+  Currencies.read()
```
This should have a `Notify` handler.

jniles commented on server/controllers/finance/reports/invoices/receipt.handlebars:
```diff
@@ -65,7 +65,7 @@
         <td>{{this.code}}</td>
         <td>{{this.text}}</td>
-        <td class="text-right">{{currency this.transaction_price}}</td>
+        <td class="text-right">{{this.transaction_price}}</td>
```
Is this for precision reasons?

sfount replied:
This is to address https://github.com/IMA-WorldHealth/bhima-2.X/issues/888

jniles commented on server/controllers/finance/reports/invoices/receipt.handlebars:
```diff
@@ -129,6 +129,22 @@
+            <td colspan="2">{{{translate 'EXCHANGE.INVOICE_DISCLAIMER'}}}</td>
+          </tr>
+          <tr>
+            <td colspan="2">{{{translate 'EXCHANGE.SET_AS'}}} 1:{{exchange}} {{{translate 'EXCHANGE.ON'}}} {{dateFormat}}</td>
```
It would look nicer if you used the currency filter on the 1 here, I think.

jniles reviewed (commented):
This looks good to me.  It's a little confusing because the currency button is in a different context from the receipt.  In the future, should we put this button on the receipt modal itself?

It's a step in the right direction for sure.

sfount committed:
Add infrasructure for receipt currency component

This commit sets up the infrastructure for the receipt currency
component. This component will provide an on update callback that
updates the controller when the user has selected a new currency for
the receipt. It will also cache the user's previous choice.

jniles commented on client/src/js/components/bhReceiptCurrency.js:
```diff
@@ -0,0 +1,48 @@
+  }
+});
+
+ReceiptCurrencyController.$inject = ['CurrencyService', 'SessionService', 'AppCache', 'Store'];
```
Nice and clean component. :+1:

jniles commented on client/src/partials/patient_invoice/registry/registry.html:
```diff
@@ -20,6 +20,12 @@
+
+
+      <div class="toolbar-item">
+        <bh-receipt-currency
```
It's better practice to have close component tags, instead of having self-closing tags.  On Chrome, our target, it isn't a problem ... but it's bad practice.

For reference, https://github.com/angular/angular.js/issues/1953, https://github.com/angular/angular/issues/5563#issuecomment-161826319.

jniles reviewed (approved):
LGTM!
